csharp/src/Ice/IceDiscovery/Locator.cs

```diff
@@ -12,229 +12,479 @@
         {
-            // Ignored
+            Debug.Assert(_replicaReplyCollector != null);
+
+            // This method is called by InvokeAsync after the first reply from a replica group to wait for additional
+            // replies from the replica group.
+            TimeSpan latency = (Time.Elapsed - start) * latencyMultiplier;
+            if (latency == TimeSpan.Zero)
+            {
+                latency = TimeSpan.FromMilliseconds(1);
+            }
+            await System.Threading.Tasks.Task.Delay(latency);
+
+            SetResult(_replicaReplyCollector());
+            return await Task;
         }
+    }
 
-        public void UnregisterAdapterEndpoints(
+    /// <summary>Servant class that implements the Slice interface LookupReply.</summary>
+    internal class LookupReply : ILookupReply, IDisposable
+    {
+        internal ReplyHandler<IObjectPrx?> ReplyHandler { get; }
+
+        private readonly object _mutex = new object();
+        private readonly HashSet<IObjectPrx> _proxies = new HashSet<IObjectPrx>();
+
+        public void Dispose() => ReplyHandler.Dispose();
+
+        public void FoundObjectById(Identity id, IObjectPrx proxy, Current current, CancellationToken cancel) =>
+            ReplyHandler.SetResult(proxy);
+
+        public void FoundAdapterById(
             string adapterId,
-            string replicaGroupId,
-            Protocol protocol,
+            IObjectPrx proxy,
+            bool isReplicaGroup,
             Current current,
             CancellationToken cancel)
         {
-            lock (_mutex)
+            if (isReplicaGroup)
             {
-                _adapters.Remove((adapterId, protocol));
-
-                if (replicaGroupId.Length > 0)
+                lock (_mutex)
                 {
-                    if (_replicaGroups.TryGetValue((replicaGroupId, protocol), out HashSet<string>? adapterIds))
+                    _proxies.Add(proxy);
+                    if (_proxies.Count == 1)
                     {
-                        adapterIds.Remove(adapterId);
-                        if (adapterIds.Count == 0)
-                        {
-                            _replicaGroups.Remove((replicaGroupId, protocol));
-                        }
+                        // Cancel WhenAny and let InvokeAsync wait for additional replies from the replica group, and
+                        // later call CollectReplicaReplies.
+                        ReplyHandler.Cancel();
                     }
                 }
             }
+            else
+            {
+                ReplyHandler.SetResult(proxy);
+            }
         }
 
-        internal (IObjectPrx? Proxy, bool IsReplicaGroup) FindAdapter(string adapterId, Protocol protocol)
+        internal LookupReply() =>
+            ReplyHandler = new ReplyHandler<IObjectPrx?>(this, () => null, CollectReplicaReplies);
```

Comment:
It could just be `new (...)` like other calls.

Reply:
Fixed.